src/zkevm_specs/evm/util/call_gadget.py

```diff
@@ -0,0 +1,127 @@
+from typing import Tuple
+from zkevm_specs.evm.table import AccountFieldTag
+from zkevm_specs.util.arithmetic import RLC
+from zkevm_specs.util.hash import EMPTY_CODE_HASH
+from zkevm_specs.util.param import (
+    GAS_COST_ACCOUNT_COLD_ACCESS,
+    GAS_COST_CALL_WITH_VALUE,
+    GAS_COST_NEW_ACCOUNT,
+    GAS_COST_WARM_ACCESS,
+)
+from ...util import (
+    FQ,
+    N_BYTES_ACCOUNT_ADDRESS,
+    N_BYTES_GAS,
+    RLC,
+)
+from ..instruction import Instruction
+
+
+class CallGadget:
+    is_oog: FQ
+
+    is_success: RLC
+    value: RLC
+
+    gas: FQ
+    callee_address: FQ
+    cd_offset: FQ
+    cd_length: FQ
+    rd_offset: FQ
+    rd_length: FQ
+
+    next_memory_size: FQ
+    memory_expansion_gas_cost: FQ
+
+    def __init__(self, instruction: Instruction, is_oog: FQ, is_call_or_callcode: FQ = FQ(0)):
```

Comment:
Instead of a class, is it possible to make the common part of callop and oog_call a function? 

For example,

```
def call_common(instruction: Instruction):
    ...
```
It would be okay to define a struct to hold the output value.

```
class CallStruct:
    is_oog: FQ
    is_success: RLC
    value: RLC
    gas: FQ
    callee_address: FQ
    cd_offset: FQ
    cd_length: FQ
    rd_offset: FQ
    rd_length: FQ
    next_memory_size: FQ
    memory_expansion_gas_cost: FQ
```

The reason is that the Python code serves as the specification of the circuit. Organizing the constraints in the class and methods makes the reader hard to keep track of variables.

Reply:
> The reason is that the Python code serves as the specification of the circuit. Organizing the constraints in the class and methods makes the reader hard to keep track of variables.

I changed back to class-base implementation to align with the implementation in `zkEVM-Circuit`([#1023](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023)). Would like to know your opinion. There are some failed cases and  I'll keep working on fixing them if you're fine with this change.

Btw, I rollback @ed255's change in #354, [here ](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/354/files#diff-10487489ead1ac341b7671be200b7c7b4b63c000a33f05ffb0bcfe63e73d6ff0R131-R133). The original way is more consistent with the implementation in `zkEVM-circuit`. Do you have any concern?